man/man1/pmlogger_rewrite.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2013-2019 Red Hat.
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.TH PMLOGGER_REWRITE 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmlogger_rewrite\f1 \- helper script to rewrite Performance Co-Pilot archives
.SH SYNOPSIS
.B $PCP_BINADM_DIR/pmlogger_rewrite
[\f3\-dNsVvw?\f1]
[\f3\-c\f1 \f2config\f1]
\f2archive\f1 ...
.SH DESCRIPTION
.B pmlogger_rewrite
is a helper script that is used by
.BR pmlogger_daily (1)
when rewriting Performance Co-Pilot (PCP)
archives.
Rewriting is usually required to accommodate evolutionary changes in metadata
so old archives can be updated and then merged with current archives.
.PP
.B pmlogger_rewrite
is a wrapper script for
.BR pmlogrewrite (1)
that will potentially rewrite all of the archives matching the
.I archive
arguments.
Each
.I archive
argument may be a PCP archive file name, the basename for
the files in a PCP archive, or a directory (in which
case all subordinate PCP archives are found by recursive
descent).
.PP
.B pmlogrewrite
is run with the
.B \-i
option so rewriting is done ``in place''.
The
.B \-q
option is also used, so if no rewriting is required then no changes are made
to the archive files.
.SH OPTIONS
The following options are passed through to
.BR pmlogrewrite :
.hy 0
.B \-c
(or
.BR \-\-config ),
.B \-d
(or
.BR \-\-desperate ),
.B \-s
(or
.BR \-\-scale ),
.B \-v
(or
.B \-\-verbose)
and
.B \-w
(or
.BR \-\-warnings ).
Refer to
.BR pmlogrewrite (1)
for the meaning of these options.
.br
.hy
.PP
The remaining options are:
.TP 5
\fB\-N\fR, \fB\-\-showme\fR
Runs
.B pmlogger_rewrite
in a ``show me'' or ``dry run'' mode where the
tasks that would be performed are reported, but no changes are made.
This is typically used for debugging in combination with one (verbose)
or two (very verbose)
.B \-V
options.
.TP
\fB\-V\fR, \fB\-\-verbose\fR
Increase script output verbosity.
.TP
\fB\-?\fR, \fB\-\-help\fR
Display usage message and exit.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmlogger (1),
.BR pmlogger_daily (1)
and
.BR pmlogrewrite (1).


CHANGELOG
pcp-6.2.2 (29 May 2024)
    - Work-in-progress, not yet released.

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging
    - Server-side utilities and log management scripts:
	pmlogrewrite, pmlogredact: fix for missing data volumes
	pmcd: cosmetic change to avoid pipe2 name clash
	runaspcp: new helper utility for unprivileged rc scripts
	pmnsmerge: be more defensive in qsort comparison routine
	pmlogsize: fix qsort comparison function handling of equality
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: pmOpenLog refactor for handling previous log files
	libpcp: new __pmCleanMapDir routine to aid log culling
	libpcp & pmcd: READY-NOT-READY hardening
	libpcp: avoid possible thread issues (coverity)
	libpcp: refactor pmDupContext for thread safety
	libpcp_import: add guard in pmiPutMark to prevent SIGSEGV
    - Build, infrastructure and packaging updates:
	build: changes to re-enable builds for FreeBSD 12, 13, 14
	build: add diffutils rpm dep as pmdaproc script uses diff
	build: add main.yml for building and publishing container
	build: resolve compiler warnings in various tools/libraries
	packaging: use tmpfiles.d method to create PCP_RUN_DIR
	build: packaging changes for openSUSE 15.5 and 15.6
	debian: do not ship empty directory /usr/lib/pkgconfig
	build: el10 systemd-presets for pmcd/pmie/pmlogger enabling
	build: drop rpm dep on bpftool as local version used only
	build: change perl YAML::XS::LibYAML module to YAML::XS
    - Security Enhanced Linux:
	selinux: policy improvements for CentOS/RHEL 7 and 8
    - Documentation:
	docs: Fedora 39 found some tbl(1)/troff(1) issues, fix
	docs: add an explicit code-of-conduct for PCP
	docs: pmns(5) note that metric names must only contain ASCII
	docs: update the BPF README to describe bpftool reality

pcp-6.1.1 (17 November 2023)
    - Client tools and utilities:
	pmie: add -o/--format to control output format with archives

src/pmlogger/pmlogger_rewrite.sh
#! /bin/sh
#
# Copyright (c) 2014 Red Hat.
# Copyright (c) 1995,2003 Silicon Graphics, Inc.  All Rights Reserved.
# Copyright (c) 2018 Ken McDonell.  All Rights Reserved.
# 
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# wrapper for pmlogrewrite
#
# exit status
# 0	nothing changed
# 1	at least one PCP archive was rewritten
# 2	non-fatal warning
# 4	aborted
#

# Get standard environment
. $PCP_DIR/etc/pcp.env
# ... and _is_archive()
. $PCP_SHARE_DIR/lib/utilproc.sh

prog=`basename $0`
tmp=`mktemp -d "$PCP_TMPFILE_DIR/pmlogger_rewrite.XXXXXXXXX"` || exit 1
status=0
trap "rm -rf $tmp; exit \$status" 0 1 2 3 15

cat > $tmp/usage << EOF
# Usage: [options] archive ...

Options:
  -c=FILE,--config=FILE  [pmlogrewrite] config file/dir
  -d, --desperate        [pmlogrewrite] desperate, save output archive even after error
  -N, --showme           perform a dry run, showing what would be done
  -s, --scale            [pmlogrewrite] do scale conversion
  -V, --verbose          increase diagnostic verbosity
  -v                     [pmlogrewrite] increased diagnostic verbosity
  -w, --warnings         [pmlogrewrite] emit warnings
  --help
EOF

# option parsing
ARGS=`pmgetopt --progname=$prog --config=$tmp/usage -- "$@"`
if [ $? != 0 ]
then
    status=4
    exit
fi

verbose=false
very_verbose=false
showme=false
rewrite_args="-iq"
eval set -- "$ARGS"
while [ $# -gt 0 ]
do
    case "$1"
    in

	-c)	shift
		rewrite_args="$rewrite_args -c \"$1\""
		;;

	-d)	rewrite_args="$rewrite_args -d"
		;;

	-N)	showme=true
		;;

	-s)	rewrite_args="$rewrite_args -s"
		;;

	-V)	if $verbose
		then
		    very_verbose=true
		else
		    verbose=true
		fi
		;;

	-v)	rewrite_args="$rewrite_args -v"
		;;

	-w)	rewrite_args="$rewrite_args -w"
		;;

	--)	shift
		break
		;;

	-\?)	pmgetopt --usage --progname=$prog --config=$tmp/usage
		status=4
		exit
		;;
    esac
    shift
done

if [ $# -eq 0 ]
then
    pmgetopt --usage --progname=$prog --config=$tmp/usage
    status=4
    exit
fi

# get $compress_suffixes variable
eval `pmconfig -L -s compress_suffixes`

# For each command line argument, if it is a directory descent to find
# all files and insert these into the command line arguments list in
# place of the directory name.
# Then for each (file) name argument, reduce it to a PCP archive base
# name (strip compression suffix, strip .index or .meta or the
# volume number), check the corresponding metadata file exists and this
# is not a base name we've already seen ... if you get this far, add the
# base name to the list of archives to be processed
#
rm -f $tmp/ags
touch $tmp/args
for try
do
    if [ -d "$try" ]
    then
	# crude filter here ... more precise filtering later on
	#
	find "$try" -type f \
	| grep -E '(\.meta|\.index|\.[0-9][0-9]*)($|(\.xz|\.lzma|\.bz2|\.bz|\.gz|\.Z|\.z)$)' >>$tmp/args
    else
	echo "$try" >>$tmp/args
    fi
done
rm -f $tmp/archives
touch $tmp/archives
for try in `cat $tmp/args`
do
    base="$try"
    for suff in $compress_suffixes
    do
	if echo "$try" | grep "$suff\$" >/dev/null
	then
	    base=`echo "$try" | sed -e "s/$suff\$//"`
	    break
	fi
    done
    # only one valid match is possible below ...
    #
    case "$base"
    in
	*.index)
	    base=`echo "$base" | sed -e 's/\.index$//'`
	    ;;
	*.meta)
	    base=`echo "$base" | sed -e 's/\.meta$//'`
	    ;;
	*.[0-9]|*.[0-9][0-9]|*.[0-9]*[0-9])
	    base=`echo "$base" | sed -e 's/\.[0-9]*[0-9]$//'`
	    ;;
	*)
	    # may already be the base name of an archive ... look for
	    # a .meta or .meta.* file
	    #
	    rm -f $tmp/ok
	    if [ -f "$base".meta ]
	    then
		touch $tmp/ok
	    else
		for suff in $compress_suffixes
		do
		    if [ -f "$base.meta"$suff ]
		    then
			touch $tmp/ok
			break
		    fi
		done
	    fi
	    if [ ! -f $tmp/ok ]
	    then
		base=''
		$verbose && echo "Warning: $try: not a PCP archive name"
	    fi
	    ;;
    esac
    [ -z "$base" ] && continue
    if [ "$try" != "$base" ]
    then
	$very_verbose && echo "suffix stripping: arg: $try -> $base"
    fi
    if grep "^$base" $tmp/archives >/dev/null
    then
	$very_verbose && echo "$base: duplicate, already in archives list"
    else
	check=`echo $base.meta*`
	if $showme || _is_archive "$check"
	then
	    echo "$base" >>$tmp/archives
	fi
    fi
done

if [ ! -s $tmp/archives ]
then
    $verbose && echo "Warning: no PCP archives found"
    exit
fi

# now for every selected archive ... check it is not active and the archive
# is OK, then rewrite and if a file is changed, recompress if the original
# file was compressed
#
for archive in `cat $tmp/archives`
do
    if $showme
    then
	echo "+ check $archive not active and OK"
	echo "+ pmlogrewrite $rewrite_args $archive"
	continue
    fi
    pid=`pmdumplog -L $archive | sed -n -e '/^PID for pmlogger: /s///p' 2>/dev/null`
    if [ -z "$pid" ]
    then
	pmdumplog -L $archive
	echo "Botch: cannot get pmlogger PID from label record for $archive"
	continue
    fi
    if pmprobe -I pmcd.pmlogger.port | grep "\"$pid\"" >/dev/null
    then
	$verbose && echo "Warning: skip archive $archive, pmlogger PID $pid is still running"
	continue
    fi
    if pmlogcheck -w "$archive" >$tmp/out 2>&1
    then
	# OK
	:
    elif [ -s $tmp/out ]
    then
	$verbose && cat $tmp/out
	echo "Warning: $base: bad archive, rewriting skipped"
	continue
    else
	# empty output but non-zero exit status?
	#
	echo "Warning: $base: bad archive (pmlogcheck exit status=$?), rewriting skipped"
	continue
    fi
    if `which sum >/dev/null 2>&1`
    then
	SUM=sum
    elif `which cksum >/dev/null 2>&1`
    then
	SUM=cksum
    else
	# Using wc(1) is lame, but no real choice!
	#
	echo "Warning: can't find sum(1) or chksum(1)"
	SUM="wc -c"
    fi
    # use sum(1) or cksum(1) to detect changes at the level of individual files
    #
    rm -f $tmp/sum.before
    for file in $archive.*
    do
	echo "$file `$SUM "$file" | sed -e 's/[ 	][ 	]*/ /g'`" >>$tmp/sum.before
    done

    eval pmlogrewrite $rewrite_args "$archive"

    rm -f $tmp/sum.after
    for file in $archive.*
    do
	echo "$file `$SUM "$file" | sed -e 's/[ 	][ 	]*/ /g'`" >>$tmp/sum.after
    done
    # now use comm to find the lines that are in $tmp/sum.after and not
    # in $tmp/sum.before ... these are the files that have been changed
    # by pmlogrewrite
    #
    comm -13 $tmp/sum.before $tmp/sum.after \
    | while read file stuff
    do
	old_file=`grep "^$file" $tmp/sum.before | sed -e 's/[ 	].*//'`
	if [ "$file" != "$old_file" ]
	then
	    # original file name and current file name do not match ...
	    # only cause should be compressed input file, uncompressed
	    # output file
	    #
	    case "$old_file"
	    in
		*.xz|*.lzma)
		    if which xz >/dev/null 2>&1
		    then
			# same logic as in pmlogger_daily for optimal
			# compression
			#
			if xz -0 --block-size=10MiB </dev/null >/dev/null 2>&1
			then
			    # want minimal overheads, -0 is the same as --fast
			    xz -0 --block-size=10MiB "$file"
			else
			    xz "$file"
			fi
			$very_verbose && echo "changed and recompressed: $old_file"
		    else
			echo "Warning: no xz(1), cannot recompress $file"
		    fi
		    ;;
		*.bz2|*.bz)
		    if which bzip2 >/dev/null 2>&1
		    then
			bzip2 "$file"
			$very_verbose && echo "changed and recompressed: $old_file"
		    else
			echo "Warning: no bzip2(1), cannot recompress $file"
		    fi
		    ;;
		*.gz|*.Z|*.z)
		    if which gzip >/dev/null 2>&1
		    then
			gzip "$file"
			$very_verbose && echo "changed and recompressed: $old_file"
		    else
			echo "Warning: no gzip(1), cannot recompress $file"
		    fi
		    ;;
		*)
		    echo "Botch: cannot handle rewriting file name change: $old_file -> $file"
		    ;;
	    esac
	else
	    $very_verbose && echo "changed: $old_file"
	fi
    done
done

exit
